Prometheus: preserve job/instance when translating Prometheus -> OTLP (#4956)
open-telemetry/opentelemetry-specification: specification/compatibility/prometheus_and_openmetrics.md

**Status**: [Development](../document-status.md)

Prometheus identifies the origin of metrics by [`job` and `instance` labels](https://prometheus.io/docs/concepts/jobs_instances/).
These labels are added when the metrics are scraped from the Prometheus endpoint,
where the `job` is set based on the scrape configuration, and the `instance` is
set to the target's address. Even though these are added at scrape time, they
are considered identifying for metrics regardless of whether they are scraped or
pushed (using Prometheus Remote Write).

Prometheus also stores metadata associated with scraped targets in the
[target_info](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)
Metric (which belongs to the `target` MetricFamily in OM 1.0).

`job` and `instance` labels SHOULD be stored as `prometheus.job` and
`prometheus.instance` resource attributes, respectively.

> If present, the `target_info` metric MUST be dropped from the batch of metrics,

Reviewer: Confused about where do you refer that this `target_info` metric is, in what "batch"?

Author: In the context of Prometheus -> OTLP, the batch is is a prometheus remote write request, or a single scrape of a prometheus endpoint.

Author: Clarified in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/4dda69a6fb71e13e25f93fde75fa378f0e9986db

and all labels from the `target_info` metric MUST be converted to resource
attributes attached to all other metrics which are part of the scrape. By
default, label keys and values MUST NOT be altered (such as replacing `_` with
`.` characters in keys).

If `service.name` or `service.instance.id` are not provided by `target_info`,

> they MUST be defaulted to the `job` and `instance` labels, respectively.

Reviewer 2: I wonder if `job` has the form of `namespace/name` whether it should be used to set `service.namespace` accordingly. I was trying to square this with the text on lines 640-642, which would insert the service.namespace, and then if there was another level of federation the service.namespace value might disappear.

Author: `job` would be formed that way if you are using a prometheus or PRW collector exporter (since SDKs don't populate job/instance themselves), and then need to translate it back to OTLP at a future point. Generally, that isn't a thing that happens much in practice, since it is much easier to just send OTLP than to round-trip through PRW or the Prometheus format. So it is something we could do, but probably has little value in practice.

Overall, I'm trying to deemphasize `job == service.name` in this PR to allow them to be separate things conceptually.

> they MUST be defaulted to the `job` and `instance` labels, respectively.

Reviewer: What if in the batch there are metrics with different job/instance pair?

Author: That would mean that the batch of Prometheus metrics produces a batch of OTel metrics with multiple resources, and metric data points nested under the associated resource.

Author: I clarified this in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/4dda69a6fb71e13e25f93fde75fa378f0e9986db

#### Resource Attributes for Scraped Targets

When Prometheus metrics are scraped from a Prometheus endpoint, the following
attributes SHOULD be associated with scraped metrics as resource attributes,
and MUST NOT be added as metric attributes:

| OTLP Resource Attribute | Description |
| ----------------------- | ----------- |
| `server.address` | The `<host>` portion of the target's URL that was scraped |
| `server.port` | The `<port>` portion of the target's URL that was scraped |

> | `url.scheme` | `http` or `https` |

Reviewer: Should we not put job/instance as the service.name/service.instance.id?

Author: Addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/4dda69a6fb71e13e25f93fde75fa378f0e9986db, which now says that these are in addition to the attributes above.

If the target was discovered using service discovery, metadata about the
discovered target SHOULD be added as resource attributes, using OpenTelemetry

> semantic conventions where possible.

Reviewer 3: How realistic is this? Does Prometheus SD metadata comply with OTel semantic conventions? 

And how can we implement this when several semantic conventions are still unstable?

Author: We already do this for k8s: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/b9a10ff7c2533cc94737600a5850b20aeb2e183b/receiver/prometheusreceiver/internal/prom_to_otlp.go#L58

We will need to put this behind a feature gate for any conventions that are not stable.  They are mostly marked beta right now: https://opentelemetry.io/docs/specs/semconv/resource/k8s/

Any that aren't stable will need to be behind a feature flag

Reviewer 3: Gotcha, I like the feature gate approach.

We should have tests to ensure we catch behavior changes on the Prometheus side. I haven't looked at the code, so maybe we already have them, but I just want to prevent us getting caught by surprise during a dependency update

Prometheus Remote Write request), there is one OpenTelemetry Resource that
contains all metrics with that `job` and `instance`. `job` and `instance` labels
are added as resource attributes, and not as metric attributes.

Prometheus also stores metadata associated with scraped targets in the
[target_info](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)
Metric (which belongs to the `target` MetricFamily in OM 1.0).

If present, all `target_info` metrics MUST be dropped and their labels
processed as follows: All labels from each `target_info` metric, except `job`
and `instance`, MUST be converted to resource attributes on the resource
associated with that `job` and `instance`. By default, all other label keys and
values MUST NOT be altered (such as replacing `_` with `.` characters in keys).
If not included as labels on `target_info`, `service.name` and
`service.instance.id` MAY default to the `job` and `instance` values, respectively.

| OTLP Resource Attribute | Description |
| ----------------------- | ----------- |

> | `service.name` | The value of the `service.name` label on `target_info`. Otherwise, MAY default to `job`. |

Reviewer 4: The prometheusremotewritereceiver also [splits](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/6004654ac94df40736b67619e00ba7320c2305a1/receiver/prometheusremotewritereceiver/receiver.go#L663) the job along `/` to get `service.namespace.`

Author: I would prefer if we didn't do that. Service names can contain slashes, and I would rather do the simpler thing that is more correct.

Reviewer 5: Not related to @krajorama's comment.

but to the statement that `service.name` will be looked at in `target_info`, ie the `.` label name rather than the `_`version and in `target_info

Did we check the developer experience when using a Prometheus client lib to use a `.` label name instead of the Prometheus traditional `_` and to set this label on `target_info` rather than in the metric itself.

Below is a check with the [Prometheus Java Metrics Library](https://github.com/prometheus/client_java) and with the [Spring Boot Actuator Prometheus](https://docs.spring.io/spring-boot/reference/actuator/metrics.html#actuator.metrics.export.prometheus) integration?

* For the Prometheus client lib to expose metric label names with `.` instead of `_` , the scrapper (eg the otelcol `prometheus` receiver) must pass the HTTP header `escaping=allow-utf-8` according to the Prometheus docs [UTF-8 metric and label name escaping schemes](https://github.com/prometheus/docs/blob/main/docs/instrumenting/escaping_schemes.md)
* For the Prometheus client lib to produce the `target_info` metric, I guess the developer will manually create this metric.
   * With the [Prometheus Java Metrics Library](https://github.com/prometheus/client_java), I guess it would look like:

      ```java
      Info targetInfo = Info.builder()
         .name("target_info")
         .labelNames(
           "service.name", 
           "service.namespace", 
           "service.instance.id",
           "service.version")
         .register();
      
      targetInfo.setLabelValues(
        "fraud-detection", 
        "astroshop", 
        "instance-123456", 
        "1.0.0");
      ```
   * With Spring Boot Actuator, a quick search on Claude code suggests it will look like:
      ```java
      return registry -> Gauge.builder("target_info", () -> 1)
              .description("Target metadata")
              .tag("service.name", "fraud-detection")
              .tag("service.namespace", "astroshop")
              .tag("service.instance.id", "instance-123456")
              .tag("service.version", version)
              .register(registry);
      ```

The code requires some expertise but it looks OK to me.

Author: @cyrille-leclerc is that related to @krajorama's comment? I haven't checked those, but you are welcome to do so.

Reviewer 5: Sorry for the confusion, my comment was not related to @krajorama 's point but to the sentence itself. I updated my comment.
I wanted to check that the developer experience looks good, particularly to define labels with `.`instead of `_` and to produce a `target_info` metric.

The code snippets I end up with above look ok to me. I didn't check for other programming languages but I guess it will be similar.

> | `service.instance.id` | The value of the `service.instance.id` label on `target_info`. Otherwise, MAY default to `instance`. |

Reviewer 5: Shall we call out that these sentences imply that the prometheus2otel conversion, ie the OTel Collector `prometheus` receiver, should have a configuration option to enable/disable derivation of OTel resource attributes `service.name` & `service.instance.id` from Prometheus `job` & `instance` labels? 
Moreover, if prometheus2otel doesn't produce `service.name` & `service.instance.id`, should `job` and `instance` be captured on the OTel metric as resource attributes or as datapoint attributes?
Why not capturing `job` and `instance` as resource/datapoint attributes all the time, whatever `service.name` and `service.instance.id` are derived from `job` and `instance` or not.

Author: Addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/3f74736a8e5da0c4216e4fc50f9877b84e54042f.

The current spec also explicitly calls out (in the paragraph above) that we always preserve job and instance as resource attributes. LMK if I can make that clearer.

#### Resource Attributes for Scraped Targets

In addition to the attributes above, when Prometheus metrics are scraped from a
Prometheus endpoint, the following attributes SHOULD be associated with scraped
metrics as resource attributes, and MUST NOT be added as metric attributes:

| OTLP Resource Attribute | Description |
| ----------------------- | ----------- |
| `server.address` | The `<host>` portion of the target's URL that was scraped |

> | `server.port` | The `<port>` portion of the target's URL that was scraped |

Reviewer 5: Are there risks of conflicts? For example a MySql server may want to report `server.port=5432` and, for `server.address` it could be more relevant to let the [resource detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/resourcedetectionprocessor) deal with it

Could `prometheus.server.port` and `prometheus.server.address` help disambiguate?

ref https://opentelemetry.io/docs/specs/semconv/registry/attributes/server/

Author: Thats a good point.  Lets address that separately from this PR if thats OK.

Reviewer 5: shall we resolve this comment or keep it open until a task is tracked?

Author: https://github.com/open-telemetry/opentelemetry-specification/issues/5146

**Status**: [Development](../document-status.md)

Prometheus identifies the origin of metrics by [`job` and `instance` labels](https://prometheus.io/docs/concepts/jobs_instances/).
These labels are added when the metrics are scraped from the Prometheus endpoint,
where the `job` is set based on the scrape configuration, and the `instance` is
set to the target's address.

`job` and `instance` labels are considered identifying for the purpose of
grouping Prometheus metrics into OpenTelemetry resources. For each
combination of `job` and `instance` labels on metrics in the incoming batch of
Prometheus metrics (e.g., a single scrape of a Prometheus endpoint or a single
Prometheus Remote Write request), there is one OpenTelemetry Resource that
contains all metrics with that `job` and `instance`. `job` and `instance` are
dropped from all metrics, and are added as the `prometheus.job` and
`prometheus.instance` resource attributes, respectively.

Prometheus also stores metadata associated with scraped targets in the
[target_info](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)

> Metric (which belongs to the `target` MetricFamily in OM 1.0).

Reviewer 6: Is this correct to say? I'm only aware that Prometheus' OTLP endpoint generates `target_info` to encode OTel resource attributes. I think the [linked document](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems) instead says that an exposer may provide the `target_info` metric for Prometheus to scrape.

The following paragraph reflects my point about `target_info` containing OTel resource attributes.

Author: The intent is to say that prometheus endpoints, when they want to expose metadata associated with themselves, do so using the "target_info" metric by convention. "stores" is probably the wrong word here, as it is more like:

```suggestion
Prometheus exporters also expose metadata associated with their target using the
[target_info](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)
Metric (which belongs to the `target` MetricFamily in OM 1.0).
```

If present, all `target_info` metrics MUST be dropped and their labels
processed as follows: All labels from each `target_info` metric, except `job`
and `instance`, MUST be converted to resource attributes on the resource
associated with that `job` and `instance`. By default, all other label keys and
values MUST NOT be altered (such as replacing `_` with `.` characters in keys).
If not included as labels on `target_info`, `service.name` and
`service.instance.id` default to the `job` and `instance` values, respectively.

| OTLP Resource Attribute | Description |
| ----------------------- | ----------- |

> | `service.name` | The value of the `service.name` label on `target_info`. Otherwise, defaults to `job`. |

Reviewer 7: Prometheus endpoints expose data about telemetry sources which are not always services. 

A while back I [proposed codifying that service.*](https://github.com/open-telemetry/semantic-conventions/pull/630) apply to all telemetry sources. The PR was merged, [but then reverted](https://github.com/open-telemetry/semantic-conventions/pull/638), and the [next attempt was rejected](https://github.com/open-telemetry/semantic-conventions/pull/671).

I think that adding `service.name` / `service.instance.id` attributes to all prometheus scrape targets is aligned with my proposal which was ultimately rejected. 

I know prometheus job and instance represent the identity for the scrape target. Prometheus uses these as sort of universal id. OpenTelemetry appears to want to go in a different direction, where different types of telemetry producers (or entities you might call them) use different attributes to encode their identity. 

I think that most of the time setting job and instance to service.name and service.instance.id will be correct. And that it makes sense as a default. But I also think that because OTEL rejected my proposal to treat service.* attributes as universal identifiers, the prometheus receiver needs configuration options to override which attributes job and instance get mapped to.

Something like:

- `job` is mapped to the configured job resource attribute mapping key, or `service.name` if none is set.
- `instance` is mapped to the configured instance resource attribute mapping key, or `service.instance.id` if none is set.

Reviewer 5: For Prometheus -> OTel, and especially Prometheus -> OTel -> Prometheus, I would prefer diverging from the current specification that maps `service.namespace/service.name=job` and `service.instance.id=instance`.

Instead, I would like to preserve:

* Native Prometheus labels such as:
    `job="prometheus_simple/10.42.0.15:8080"` and `instance="10.42.0.15:8080"`
    to maintain Prometheus consistency and, ideally, backward compatibility with existing Prometheus instrumentation and queries.
* Native OTel resource attributes such as `service.name`, `service.namespace`, and `service.instance.id`, generated according to [SemConv: Specify resource attributes using Kubernetes annotations](https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/)￼, so they correlate correctly with SDK telemetry and pod logs.

Concretely, I think the OTel Collector Prometheus Receiver on Kubernetes (Prometheus -> OTel) should stop deriving `service.name` and `service.instance.id` from the Prometheus job and instance labels, and instead rely on the k8sattributes processor to populate these resource attributes according to the Kubernetes semantic conventions.

Today, the OTel Collector Prometheus Receiver on Kubernetes generates resource attributes such as:

```
service.name="prometheus_simple/10.42.0.15:8080"
service.instance.id="10.42.0.15:8080"
```

This is misaligned with the Kubernetes semantic conventions and prevents proper correlation with pod logs and telemetry produced by OTel SDK instrumentation.

I think it relates to our finding with @aknuds1 and @lmolkova considering generating `service.name` and `service.instance.id` in the OTelCol Postgresql receiver: it doesn't work well on k8S as it's misaligned with [SemConv: Specify resource attributes using Kubernetes annotations](https://opentelemetry.io/docs/specs/semconv/non-normative/k8s-attributes/) and thus breaks correlation with pod logs...

Author: > I think that most of the time setting job and instance to service.name and service.instance.id will be correct. And that it makes sense as a default. But I also think that because OTEL rejected my proposal to treat service.* attributes as universal identifiers, the prometheus receiver needs configuration options to override which attributes job and instance get mapped to.

I was thinking of simpler configuration to just allow NOT defaulting service.* resource attributes to job/instance. Users can use processors to use the mapping they want in a more flexible way than we can support in the receiver. To @cyrille-leclerc's point above, we could think about going one step further, and changing the default to not set them at all, and allow opting-in to setting service.* from job/instance.

Given how widely adopted the Prometheus receiver is, I would rather not change the default behavior at this point, as that would likely break a lot of users. cc @ArthurSens for another opinion.

> Native Prometheus labels such as:
job="prometheus_simple/10.42.0.15:8080" and instance="10.42.0.15:8080"
to maintain Prometheus consistency and, ideally, backward compatibility with existing Prometheus instrumentation and queries.

@cyrille-leclerc That is the direction i'm trying to go with this proposal, but OpenTelemetry uses namespaces for all attributes to prevent collisions, so i've opted to use `prometheus.job` and `prometheus.instance` instead while it is in an OpenTelemetry format.

Reviewer 5: >  just allow NOT defaulting service.* resource attributes to job/instance

That's music to my ears.

Note that we could make this the default behaviour on OpenTelemetry Helm Charts presets.

See:
- https://github.com/open-telemetry/opentelemetry-helm-charts/pull/2206#issuecomment-4487438896

Reviewer 7: > I was thinking of simpler configuration to just allow NOT defaulting service.* resource attributes to job/instance. Users can use processors to use the mapping they want in a more flexible way than we can support in the receiver. 

That's a good point and a good idea. I support.

>  would rather not change the default behavior at this point

Just to be explicit, the collector currently does job / instance to service.name / service.instance.id by default and currently there is no receiver configuration to change this behavior, yeah?

Author: Yes.  Today, the collector simply renames job -> service.name and instance -> service.instance.id unconditionally, although the service.name and service.instance.id labels from target_info takes precedence if that is provided (and you would lose your job/instance!).

Reviewer 3: I'm on board with this configuration option to not transform job/instance into `service.*` resource attributes. 

The whole end-to-end flow needs to be properly documented, though. Not here, of course, but I doubt users will have an easy time understanding all the processors that need to be run alongside the Prometheus receiver for them to get a good experience in an OTLP-focused backend

Reviewer 5: > The whole end-to-end flow needs to be properly documented,

+100% I think that the otel collector and kube stack helm charts, alongside with the otel-demo, are great places to test, validate, and provide the first implementation of such a setup. 
I'll be happy to help on this

Reviewer 5: FYI I submitted a PR on the demo to showcase integration in OTel of Prometheus based application custom metrics (scrapping with the OTel Collector, correlation...):
- https://github.com/open-telemetry/opentelemetry-demo/pull/3388

Reviewer 5: FYI enhancement of the OTel Demo to showcase Prometheus custom application metrics integration in an OTel ecosystem:
- https://github.com/open-telemetry/opentelemetry-demo/pull/3388

Author: @ArthurSens WDYT about `prometheus.job` vs `job`?  At first, I was pretty convinced that we need to namespace the label.  But OTOH, we get a lot of non-namespaced labels from Prometheus (pretty much no labels are namespaced). Maybe we shouldn't treat job/instance as special here, and should just pass them through as-is. Prometheus already has `honor_labels` to resolve label conflicts for job/instance anyways, so maybe that isn't a concern?  I think it will reduce user confusion.

Author: I've gone ahead and made that change.  I'll resolve this thread, as I think most concerns are addressed.

> 

Author: Stale branch: CHANGELOG conflicts with main and `make markdown-toc-check` fails after rebase. merge main, resolve the CHANGELOG by moving the entry to main's current.

**Status**: [Development](../document-status.md)

Prometheus identifies the origin of metrics by [`job` and `instance` labels](https://prometheus.io/docs/concepts/jobs_instances/).
These labels are added when the metrics are scraped from the Prometheus endpoint,
where the `job` is set based on the scrape configuration, and the `instance` is
set to the target's address.

`job` and `instance` labels are considered identifying for the purpose of
grouping Prometheus metrics into OpenTelemetry resources. For each
combination of `job` and `instance` labels on metrics in the incoming batch of
Prometheus metrics (e.g., a single scrape of a Prometheus endpoint or a single
Prometheus Remote Write request), there is one OpenTelemetry Resource that
contains all metrics with that `job` and `instance`. `job` and `instance` labels

> MUST be added as resource attributes, and not as metric attributes.

Reviewer 6: Are you sure it's a good idea to add `job` and `instance` labels as resource attributes? OTLP -> Prometheus translation considers (according to spec and implementation) instead `service.instance.id` and `service.name` to be the identfiying resource attributes. Also, I think that `job` and `instance` resource attributes would be dropped in favour of corresponding labels synthesized by Prometheus when generating `target_info`.

Author: See this discussion above: https://github.com/open-telemetry/opentelemetry-specification/pull/4956#discussion_r3250215650.  While our specification stated that service.* attributes were identifying, that was based on a misread of the service attributes specification.

This is now proposing that job and instance are respected if they already exist as resource attributes when ingesting OTLP (since it means that the metric was originally scraped from a prometheus endpoint.

Reviewer 6: @dashpole What exactly is based on a misread? That Prometheus' OTLP endpoint treats `service.instance.id` and `service.name` resource attributes as identifying? If so, how precisely? Changing it would be backwards incompatible, and I would argue outside of this PR's scope.

What about the second problem I raised, i.e. that `job` and `instance` resource attributes would be dropped by the Prometheus OTLP endpoint in favour of corresponding labels synthesized from identifying resource attributes when generating the `target_info` metric?

> This is now proposing that job and instance are respected if they already exist as resource attributes when ingesting OTLP

Where is that stated in the PR's proposed text?

Reviewer 6: > since it means that the metric was originally scraped from a prometheus endpoint

@dashpole We can't make that assumption. It's but _one_ possible source of `job` and `instance` OTel resource attributes.

Author: > We can't make that assumption. It's but one possible source of job and instance OTel resource attributes.

Fair enough. It likely came from a prometheus endpoint, but obviously isn't guaranteed to have.  But that isn't really relevant.

> What about the second problem I raised, i.e. that job and instance resource attributes would be dropped by the Prometheus OTLP endpoint in favour of corresponding labels synthesized from identifying resource attributes when generating the target_info metric?

I'm not quite sure what you mean. Where in this spec does "labels synthesized from identifying resource attributes when generating the target_info metric" come from?  Are you just referring to the fact that we currently derive job/instance from service.name/service.namespace/service.instance.id?

Reviewer 6: > Where in this spec does "labels synthesized from identifying resource attributes when generating the target_info metric" come from?

@dashpole In addition to how it currently works in Prometheus' OTLP endpoint, it's effectively stipulated in the PR's changes:

> If `job` is not present as a resource attribute and `service.name` is present,
the `service.name` and `service.namespace` resource attributes MUST be combined
as `<service.namespace>/<service.name>`, or `<service.name>` if namespace is
empty, to form the `job` metric label; otherwise, `job` SHOULD be added with an
empty value. If `instance` is not present as a resource attribute and
`service.instance.id` is present, `service.instance.id` MUST be converted to the
`instance` label; otherwise, `instance` SHOULD be added with an empty value. 
>
> The resulting `job` and `instance` labels MUST be added to all OpenTelemetry
metrics that are associated with the Resource, including the `target_info` metric,
if present.

except for the `job` and `instance` labels described under
[Aggregated Exporters](#aggregated-exporters).

If info-typed metric families are not yet supported by the language Prometheus
client library, a gauge-typed metric family named `target_info` with a constant
value of 1 MUST be used instead.

To convert OTLP resource attributes to Prometheus labels, string Attribute
values are converted directly to labels, and non-string Attribute values MUST be
converted to string attributes following the [attribute
specification](../common/README.md#attribute).

#### Aggregated Exporters

Unlike application-level Prometheus exporters, exporters which expose aggregated
metrics from multiple applications are expected to include
[`job` and `instance`](https://prometheus.io/docs/concepts/jobs_instances/)
labels to distinguish the source of the metrics. This is the case for both
["federated"](https://prometheus.io/docs/prometheus/latest/federation/) pull
metric endpoints, such as the Collector's Prometheus exporter, and Prometheus
push metric exporters, such as the Collector's
[Prometheus Remote Write](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/)
exporter.

> If `job` is not present, the `service.name`

Reviewer 5: Shall we indicate if `job` and `instance` attributes on OTel metrics are expected to be resource attributes or datapoint attributes?

Author: Addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/3f74736a8e5da0c4216e4fc50f9877b84e54042f

and `service.namespace` attributes MUST be combined as
`<service.namespace>/<service.name>`, or `<service.name>` if namespace is empty,
to form the `job` metric label. If `instance` is not present,

> the `service.instance.id` attribute, if present, MUST be converted to the

Reviewer 5: Shall we document the scenario where none of `service.name` & `job` is present? Same for `service.instance.id` & `instance`?
It's common for otel collector infra monitoring receiver to lack all of attributes above when used outside of Kubernetes.

Author: Addressed in https://github.com/open-telemetry/opentelemetry-specification/pull/4956/changes/3f74736a8e5da0c4216e4fc50f9877b84e54042f.

**Status**: [Development](../document-status.md)

OpenTelemetry Resource attributes SHOULD be converted to a

> [`target_info` Metric](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)

Reviewer 6: [Nit] Is the capitalization intentional?
```suggestion
[`target_info` metric](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)
```

Author: It is a reference to "Metric", which is defined in OpenMetrics. There isn't really a "specification" for prometheus outside of OpenMetrics, so when this spec needs to refer to a Prometheus concept, it usually references the terms defined in OpenMetrics. https://github.com/prometheus/OpenMetrics/blob/main/specification/OpenMetrics.md#metric.  In practice, it probably doesn't matter.

Reviewer 6: For reasons unknown, GitHub won't let me resolve the comment.

if the resource is not [empty](../resource/sdk.md#the-empty-resource).
The Resource attributes MUST NOT be copied to labels of exported metric families

> by default. The `target_info` Metric MUST be an info-typed metric whose labels

Reviewer 6: [Nit] Same question about capitalization.
```suggestion
by default. The `target_info` metric MUST be an info-typed metric whose labels
```

Reviewer 6: For reasons unknown, GitHub won't let me resolve the comment.

> MUST include the resource attributes, and MUST NOT include any other labels.

Reviewer 6: I think we here need to include the synthesized labels `job` and `instance`:
```suggestion
MUST include the resource attributes, and MUST NOT include any other labels
other than `job` and `instance`.
```

client library, a gauge-typed metric family named `target_info` with a constant
value of 1 MUST be used instead.

To convert OTLP resource attributes to Prometheus labels, string Attribute
values are converted directly to labels, and non-string Attribute values MUST be
converted to string attributes following the [attribute
specification](../common/README.md#attribute).

#### Aggregated Exporters

Unlike application-level Prometheus exporters, exporters which expose aggregated
metrics from multiple applications are expected to include
[`job` and `instance`](https://prometheus.io/docs/concepts/jobs_instances/)
labels to distinguish the source of the metrics. This is the case for both
["federated"](https://prometheus.io/docs/prometheus/latest/federation/) pull
metric endpoints, such as the Collector's Prometheus exporter, and Prometheus
push metric exporters, such as the Collector's
[Prometheus Remote Write](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/)
exporter.

If `prometheus.job` or `prometheus.instance` are present in the OpenTelemetry
resource attributes, they MUST be used as the `job` and `instance` labels,

> respectively, and MUST NOT be included on target_info as `prometheus.job` and

Reviewer 6: [Nit]
```suggestion
respectively, and MUST NOT be included on `target_info` as `prometheus.job` and
```

`prometheus.instance`. If `prometheus.job` is not present, the `service.name`
and `service.namespace` attributes MUST be combined as
`<service.namespace>/<service.name>`, or `<service.name>` if namespace is empty,
to form the `job` metric label. If `prometheus.instance` is not present,
the `service.instance.id` attribute, if present, MUST be converted to the
`instance` label; otherwise, `instance` should be added with an empty value.

The resulting `job` and `instance` labels MUST be added to all OpenTelemetry

> metrics associated with the Resource, including the `target_info` metric,

Reviewer 6: The following is perhaps clearer, considering Prometheus metrics are produced?
```suggestion
The resulting `job` and `instance` labels MUST be added to every Prometheus
metric produced from this Resource, including the `target_info` metric,
if present.
```

Author: This is meant to be about the structure of the data, not about the actual "entity".  I.e. OTLP has
```
Resource
    > Scope
        > Metric
            > Datapoint
```

I want to say that when we translate all of the metrics/datapoints, they get the job/instance from their own resource. Maybe:
```suggestion
The resulting `job` and `instance` labels MUST be added to all Prometheus
metrics produced from this OpenTelemetry Resource and associated Metrics,
including the `target_info` metric, if present.
```

> metrics that associated with the Resource, including the `target_info` metric,

Reviewer 6: Typo
```suggestion
metrics associated with the Resource, including the `target_info` metric,
```

> if present.

Author: ```suggestion
```

**Status**: [Development](../document-status.md)

OpenTelemetry Resource attributes SHOULD be converted to a

> [`target_info` Metric](https://github.com/prometheus/OpenMetrics/blob/v1.0.0/specification/OpenMetrics.md#supporting-target-metadata-in-both-push-based-and-pull-based-systems)

Reviewer 5: Shall we document the behavior if the OTel resource attributes lack `service.name` & `job` or `service.instance.id` & `instance`?
It's the most frequent scenario for infrastructure monitoring metrics collected by the OTel Collector when running outside of Kubernetes (e.g. OTel Collector Host Metrics or MySQL receivers).
Today the resource attributes are just dropped.

Reviewer 3: I think this is a good callout! @jack-berg tried to amend the OTLP spec so that every telemetry emitter includes service-like resource attributes, but it didn't make it into the official spec.

We should clarify what to do in these situations, but I'd suggest that we open an issue first to start the discussion. What do you think?

What I would like to avoid is blocking this PR for issues that are slightly out of scope of its objective.

Reviewer 5: +100 to NOT block the PR which, in its current states, is a great improvement. Creating an issue looks good to me.

Reviewer 5: shall we resolve this comment or keep it open until a task is tracked?

Reviewer 3: I've opened https://github.com/open-telemetry/opentelemetry-specification/issues/5145, could you double check if I wrote things correctly?

Reviewer 5: Thread resolved by https://github.com/open-telemetry/opentelemetry-specification/issues/5145
I fail to resolve this thread I opened, it's probably a permission issue. If someone can resolve it to move the PR forward, that would be great.

if the resource is not [empty](../resource/sdk.md#the-empty-resource).
The Resource attributes MUST NOT be copied to labels of exported metric families

> by default. The `target_info` Metric MUST be an info-typed metric whose labels

Reviewer 5: The phrasing "by default" is very subtle and very wise.
As we could reconfirm during the last OTel unconference sessions dedicated to OTel<->Prometheus, the promotion of resource attributes as metric labels is very popular amongst practitioners (see `promote_resource_attributes ` in Prometheus docs [Using Prometheus as your OpenTelemetry backend](https://prometheus.io/docs/guides/opentelemetry/#promoting-resource-attributes)).
Shall the spec include this resource attribute promotion mechanism?

Reviewer 3: The resource attribute promotion is already part of the spec (with a slightly different config name), but it's in the [Prometheus exporter spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk_exporters/prometheus.md#resource-attributes-as-metric-labels).

Reviewer 5: Thanks, that's perfect.

Reviewer 5: FYI I fail to resolve this comment. I don' know if it's a bug or a problem of permissions.

except for the `job` and `instance` labels described under
[Aggregated Exporters](#aggregated-exporters).

If info-typed metric families are not yet supported by the language Prometheus
client library, a gauge-typed metric family named `target_info` with a constant
value of 1 MUST be used instead.

To convert OTLP resource attributes to Prometheus labels, string Attribute
values are converted directly to labels, and non-string Attribute values MUST be
converted to string attributes following the [attribute
specification](../common/README.md#attribute).

#### Aggregated Exporters

Unlike application-level Prometheus exporters, exporters which expose aggregated
metrics from multiple applications are expected to include
[`job` and `instance`](https://prometheus.io/docs/concepts/jobs_instances/)
labels to distinguish the source of the metrics. This is the case for both
["federated"](https://prometheus.io/docs/prometheus/latest/federation/) pull
metric endpoints, such as the Collector's Prometheus exporter, and Prometheus
push metric exporters, such as the Collector's
[Prometheus Remote Write](https://prometheus.io/docs/specs/prw/remote_write_spec_2_0/)
exporter.

> If `job` is not present as a resource attribute and `service.name` is present,

Author: The reverse direction never says to use the `job`/`instance` resource attributes — the PR's entire mechanism is absent from the normative text. restructure as an explicit precedence list, and give it a name so the.

> If `job` is not present as a resource attribute and `service.name` is present,

Author: The reverse direction never says to use the `job`/`instance` resource attributes — the PR's entire mechanism is absent from the normative text

The paragraph reads: *"If `job` is not present as a resource attribute **and**

Suggested fix: restructure as an explicit precedence list, and give it a name so the

the `service.name` and `service.namespace` resource attributes MUST be combined
as `<service.namespace>/<service.name>`, or `<service.name>` if namespace is
empty, to form the `job` metric label; otherwise, `job` SHOULD be added with an
empty value. If `instance` is not present as a resource attribute and
`service.instance.id` is present, `service.instance.id` MUST be converted to the

> `instance` label; otherwise, `instance` SHOULD be added with an empty value.

Reviewer 6: On re-reading the PR, I think this must be where you intend to define that `job` and `instance` resource attributes respectively should be identifying iff present, with `service.name` and `service.instance.id` as respective fallbacks. That should be clearly spelled out, as the intent is effectively ambiguous as currently written.

There are some issues with such a radical change however:

1. Users whose resource attributes include `job` and/or `instance` (doesn't sound entirely unlikely) will experience that their series identities change, when said attributes differ from (`service.namespace`/ +)`service.name`/`service.instance.id` - or when the latter are absent.
2. The effect of the `keep_identifying_resource_attributes` config parameter becomes data-dependent: It can only meaningfully apply to the `service.name` and `service.instance.id` resource attributes synthesized into `job` and `instance`. When `job` and/or `instance` resource attributes take precedence, the former presumably have to be kept on `target_info` regardless of the parameter.

Reviewer 6: I thought about this some more in the meantime - I think that projecting the Prometheus meaning of `job` and `instance` onto OTel resource attributes is a flawed design. They don't have this intrinsic meaning within the OTel scope, so I do think that the previous proposal within this PR to prefix them with "prometheus." was an objectively better choice. See also my [comment](https://github.com/open-telemetry/opentelemetry-specification/pull/4956#discussion_r3393416850) above regarding the assumption that `job` and `instance` OTel resource attributes would be proof of Prometheus provenance.

Author: We will have to deal with how to make this change in prometheus, but I don't think it will be that difficult.  It likely just means it will be opt-in until the next prometheus major release.

I don't quite see how adding a prefix changes whether job/instance have meaning.  We copy all other labels from prometheus metrics to otel resource/metric labels without prefixes.  It isn't clear to me why these labels in particular need prefixes, but the rest of them don't.

Again, i'm open to having it prefixed, but I don't see why it would be objectively wrong to copy label keys between Prometheus and OpenTelemetry as a general rule.

Reviewer 6: @dashpole I think there's a misunderstanding around the direction here - AFAICT you are referring to Prometheus -> OTLP conversion, but what I am referring to here is the opposite direction: OTLP -> Prometheus. The specific problem that @krajorama and I are raising is that this proposal would stipulate that the Prometheus OTLP endpoint treat `instance` and `job` OTel resource attributes as identifying. We think this is a flawed design, and as a result a problem that will remain even after backwards compatibility itself is no longer required through Prometheus v4.

@krajorama and I would still like have a design doc, so the targeted use cases can be sufficiently clarified. Alternatively we could discuss the proposal during a Prometheus WG call, and produce a doc from that.

Reviewer 4: as Arve says

The resulting `job` and `instance` labels MUST be added to all OpenTelemetry
metrics that are associated with the Resource, including the `target_info` metric,
if present.

> [metricMetadata]: https://github.com/open-telemetry/opentelemetry-proto/blob/c451441d7b73f702d1647574c730daf7786f188c/opentelemetry/proto/metrics/v1/metrics.proto#L199

Author: A normative requirement is deleted with no replacement: at most one `target_info` per (job, instance). restore it verbatim into `#### Aggregated Exporters`, e.g.

> [metricMetadata]: https://github.com/open-telemetry/opentelemetry-proto/blob/c451441d7b73f702d1647574c730daf7786f188c/opentelemetry/proto/metrics/v1/metrics.proto#L199

Author: A normative requirement is deleted with no replacement: at most one `target_info` per (job, instance)

The only surviving "at most one" rule in the repo is

Suggested fix: restore it verbatim into `#### Aggregated Exporters`, e.g. after
